src/config-schemas/openmrs-esm-patient-chart-schema.ts

```diff
@@ -99,5 +101,45 @@ export const esmPatientChartSchema = {
       }
     },
     default: coreTabbedViewDefinitions
+  },
+  widgetsConfigs: {
+    dictionaryElements: {
+      skipValidations: true
+    }
```

Comment:
Oh wow, I'm realizing now that I read this that I asked you to do a bunch of totally unnecessary work. I'm really sorry, @nkimaina . I think the following would be better (putting it up with `widgetDefinitions`):

```js
widgetDefintions: {
  arrayElements: {
      name: { validators: [validators.isString] },
      esModule: { validators: [validators.isString] },
      usesSingleSpaContext: { validators: [validators.isBoolean] }
      config: { validators: [validators.isObject] }
    },
    default: coreWidgetDefinitions
  }
}
```

Reply:
No worries. I thought about it some more. Correct me if am wrong:
1. With the above approach, I would need to override a lot of default configs just to change a single config entry, like temperature concept for instance. I would need to override configs for how the widget is rendered, which module it comes from etc. As the widget definition configs changes over time, I would need to be aware of those changes and provide overrides for them as well. 
2. Some of the configs are used across multiple widgets. For example, the vitals concepts would be used by the vitals-summary widget, as well as vitals form, some external widget etc. 

In fact, we might not have done extra work, but might have provided a way to provide external configs. I would prefer to rename the 'widgetConfigs' to 'externalConfigs'. The existing widgetDefinitions are good enough as they describe how to load and render widgets. The other configs might describe how to fetch data for the widgets e.t.c. 
Generally, I think we need to prefer flat/shallow configs, and try as much as possible to avoid deeply nested configs as it just makes it complicated to patch/override 'leaf' configs, and makes them harder to read/work with. 
What are you thoughts on all this?

Reply:
>     1. With the above approach, I would need to override a lot of default configs just to change a single config entry, like temperature concept for instance. I would need to override configs for how the widget is rendered, which module it comes from etc. As the widget definition configs changes over time, I would need to be aware of those changes and provide overrides for them as well.

Yes, this is a really good observation.

>     2. Some of the configs are used across multiple widgets. For example, the vitals concepts would be used by the vitals-summary widget, as well as vitals form, some external widget etc.
> 
> In fact, we might not have done extra work, but might have provided a way to provide external configs. I would prefer to rename the 'widgetConfigs' to 'externalConfigs'.

Ooh, interesting, yes. Can we call it `widgetCommonConfig`? But it still can be like
```js
widgetCommonConfig: { validators: [isObject] }
```

an example config would be
```json
"widgetCommonConfig": {
  "temperatureConcept": "1jflaw-jeir2039-jfqa"
}
```

> The existing widgetDefinitions are good enough as they describe how to load and render widgets. The other configs might describe how to fetch data for the widgets e.t.c.

I imagine that at some point we will want widget-specific config. e.g.

```json
"widgetDefinitions": [
  {
    "name": "vitals-summary",
    ...
    "config": {
      "numEncounters": 6
    }
  }
]

> Generally, I think we need to prefer flat/shallow configs, and try as much as possible to avoid deeply nested configs as it just makes it complicated to patch/override 'leaf' configs, and makes them harder to read/work with.

Note that configs are "deep merged," so
```js
provide({ a: { b: { c: 0 }, d: { e: 0 } } })
provide({ a: { b: { c: 1 } } })
```
results in
```js
{ a: { b: { c: 1 }, d: { e: 0 } } }
```

However, config arrays are not merged in any way, so in this case your point stands. You've made a solid case for a common widget config.

In the general case, as with any denormalized structure, "flattening" requires the introduction of references (e.g. foreign keys) that add a layer of indirection and make things harder to think about. I think the normalized/nested approach should be preferred, except in cases where the same config information is needed by other parts of the config. I could imagine introducing some kind of support for array merging, as well. Once there's a really solid use case.

Reply:
sounds good. I have added the common config property. I still think it's a good idea to allow one to inject extrnal configs without being restricted by current patient chart schema. I have introduced the property 'externalPatientChartConfigs' that they can use for this purpose.

Reply:
How would that be used?

Reply:
I am trying to use it here:
https://github.com/AMPATH/ampath-esm-patient-chart-widgets/blob/4e6e81092d04c3b5ea10249a65c2f068b2b78212/src/config/programs.schema.ts#L2

and loading the config here:
https://github.com/AMPATH/ampath-esm-root-config/blob/0cc6bbe9d026ea1b7cdfafb45f275ed21a38ff8e/src/configs/programs.config.json#L1

The above configs are used by multiple widgets.

Reply:
Why is this different from `widgetCommonConfig`?

Reply:
And this word "external"... what is this config external to?

Reply:
> And this word "external"... what is this config external to?

It means it's a config that will be used by distribution specific esm-widget i.e. will not be used by either the patient-chart or patient-chart-widget. We rely on patient-chart to load these configs for us, because these 'distribution specific' widgets don't have an entry point, so it's a bit tricky to set-up. Anyways, we can remove it in this PR but this is kind of a blocker for me as I am trying to work on integrating AMPATH specific widgets right now, and I can't do it if I don't have a way to load configs for it.

Reply:
> Why is this different from `widgetCommonConfig`?

It's different because these are provided by 3rd party (patient chart widget modules created by different distributions). For e.g in the example that I shared, it shows how to configure a way to group forms by program. It's an approach that I have seen being used in Kenya, but may not be true for every other distribution/implementation in openmrs. We could add it though, if anyone else is/will be interested in using this approach.

Reply:
Right, I think you would just include those configs in the widget-specific config:
```
"widgetDefinitions": [
  {
    "name": "pacman",
    "esModule": "@ampath/esm-pacman-widget",
    "config": {
      "numGhosts": 4
    }
  }
]
```